File: templates/_server_wrapper.js
/**
 * Minimal Battlesnake server wrapper for local testing.
 * Used internally by `snake-arena test`.
 * Usage: node _server_wrapper.js <strategy.js> [port]
 */

const http = require("http");
const path = require("path");
const fs = require("fs");

const strategyPath = process.argv[2];
const port = parseInt(process.argv[3] || "8080", 10);

const code = fs.readFileSync(path.resolve(strategyPath), "utf-8");
const mod = {};
const fn = eval(`(function(module, exports) { ${code} })`);
fn(mod, mod);
const decideMove = mod.exports?.decideMove || mod.decideMove;

const server = http.createServer((req, res) => {
  if (req.method === "GET") {
    res.writeHead(200, { "Content-Type": "application/json" });
    res.end(JSON.stringify({ apiversion: "1" }));
    return;
  }
  if (req.method === "POST") {
    let body = "";
    req.on("data", (c) => (body += c));
    req.on("end", () => {
      if (req.url === "/move") {
        try {
          const data = JSON.parse(body);
          let move = decideMove(data);
          if (!["up", "down", "left", "right"].includes(move)) move = "up";
          res.writeHead(200, { "Content-Type": "application/json" });
          res.end(JSON.stringify({ move }));
        } catch {
          res.writeHead(200, { "Content-Type": "application/json" });
          res.end(JSON.stringify({ move: "up" }));
        }
      } else {
        res.writeHead(200, { "Content-Type": "application/json" });
        res.end(JSON.stringify({ ok: true }));
      }
    });
    return;
  }
  res.writeHead(404);
  res.end();
});

server.listen(port, () => {
  console.log(`Snake server on port ${port}`);
});

File: templates/_server_wrapper.py
"""Minimal Battlesnake server wrapper for local testing.
Used internally by `snake-arena test`.
Usage: python _server_wrapper.py <strategy.py> [port]
"""

import importlib.util
import json
import sys
from http.server import HTTPServer, BaseHTTPRequestHandler


class H(BaseHTTPRequestHandler):
    fn = None

    def do_GET(self):
        self.send_response(200)
        self.send_header("Content-Type", "application/json")
        self.end_headers()
        self.wfile.write(json.dumps({"apiversion": "1"}).encode())

    def do_POST(self):
        length = int(self.headers.get("Content-Length", 0))
        body = json.loads(self.rfile.read(length)) if length else {}
        if self.path == "/move":
            try:
                move = self.__class__.fn(body)
                if move not in ("up", "down", "left", "right"):
                    move = "up"
            except Exception:
                move = "up"
            resp = {"move": move}
        else:
            resp = {"ok": True}
        self.send_response(200)
        self.send_header("Content-Type", "application/json")
        self.end_headers()
        self.wfile.write(json.dumps(resp).encode())

    def log_message(self, format, *args):
        pass


spec = importlib.util.spec_from_file_location("s", sys.argv[1])
mod = importlib.util.module_from_spec(spec)
spec.loader.exec_module(mod)
H.fn = mod.decide_move
port = int(sys.argv[2]) if len(sys.argv) > 2 else 8080
server = HTTPServer(("0.0.0.0", port), H)
print(f"Snake server on port {port}", flush=True)
server.serve_forever()
